streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -342,7 +342,19 @@ public void handleAssignment(final Map<TaskId, Set<TopicPartition>> activeTasks,
 
         maybeThrowTaskExceptions(taskCloseExceptions);
 
-        createNewTasks(activeTasksToCreate, standbyTasksToCreate);
+        rebalanceInProgress = true;
+        final Collection<Task> newActiveTasks = createNewTasks(activeTasksToCreate, standbyTasksToCreate);
+        // If there are any transactions in flight and there are newly created active tasks, commit the tasks
+        // to avoid potential long restoration times.
+        if (processingMode == EXACTLY_ONCE_V2 && threadProducer().transactionInFlight() && !newActiveTasks.isEmpty()) {
```

Comment:
Why is this limited to EOS_v2? From my understanding, EOS_v1 would have the same problem?

It also only seems to be a potential issue, when we get stateful tasks assigned? Not sure if we can limit the last check to `!newActiveStatefulTasks.isEmpty()` ?

Reply:
EOS_V2 was suggested by Guozhang [here](https://github.com/apache/kafka/pull/11433#discussion_r1147064683). Yeah I think  !newActiveStatefulTasks.isEmpty() seems like the right condition to add.